### DISCUSSION TITLE
Dynamically build conditions

### DISCUSSION THREAD

**Discussion by MentiHelgi** [jeremyevans/sequel#2008] [General]
Hi. I am struggling with building (recursively) dynamic conditions (they can be nested).

I have a structure, parsed by an external library that looks something like this:
```
[ 'and', ['eq', 'field', 'value'], ['eq', 'other_field', 'other_value']]
```
The method I opted for was to do recursion, that is iterate over the array. When I meet a logical operator (and, or, not), I recurse and build the child expressions, which are then returned. This will look something like this:
```
exp = {
  "and"  => -> (expressions) { Sequel.&(expression) },
  "eq" => -> (field, value) { Sequel.lit("#{field} = ?", value) }
}
... # Array iteration
operator, *expressions = *filters
case op
when "and"
  exp[operator].call(expressions.map { |expression| parse(expression) })
when "eq"
  field, value = *expressions
  exp[operator].call(field, value)
...
```
The problem is that this only works for value operators. I can't get this to work for logical operators. I would have expected to be able to call some method like `Sequel.and` that takes an array of expressions and returns a Sequel AND object that would just translate to SQL such as: `foo = 'bar' AND baz = 'qux'`. 

Anyone have any idea what I need to do for `AND`, `OR` and `NOT` to achieve this?

**Comment by janko** [comment]
It sounds like `Sequel.&`, `Sequel.|` and `Sequel.~` are the methods you want, you can see their usage in [Sequel for SQL Users](https://sequel.jeremyevans.net/rdoc/files/doc/sql_rdoc.html#label-Boolean+Operators+-28AND+OR-29) guide.

**Reply by MentiHelgi** [reply]
For some reason, when I do that, ie. `Sequel.&([Sequel.lit(....), Sequel.lit(...)])` the resulting boolean expression is a `:NOOP` (ex: `WHERE (a = 'b', c = 'd'`). Should I be doing: `Sequel.&(*expressions)` or something like that instead?

**Reply by jeremyevans** [reply]
Yes. `Sequel.&` and `Sequel.|` take multiple arguments and combine them with `AND` or `OR`.  Passing a single argument returns the argument as-is.

**Reply by MentiHelgi** [reply]
Great. Thanks for clearing that out! :smile_cat:
